extensions/parquet/base/src/main/java/io/deephaven/parquet/base/ColumnPageReaderImpl.java

```diff
@@ -577,6 +577,15 @@ private void ensureNumValues() throws IOException {
         if (numValues >= 0) {
             return;
         }
+        if (pageHeader == null) {
+            try (final SeekableByteChannel readChannel = channelsProvider.getReadChannel(filePath)) {
+                readChannel.position(offset);
+                ensurePageHeader(readChannel);
```

Comment:
Something is wrong with the flow in this code. All of the places that `ensurePageHeader(SeekableByteChannel)` are positioning the channel to `offset`. But `offset` is non-final, and set in both the constructor and in `ensurePageHeader`. I'm hoping that overwriting it in `ensurePageHeader` is wrong and can be removed....

Reply:
The offset field is not final because it can store the offset of page header or the offset of data (which comes after the page header). So once we read the page header, we increment the offset so that following reads directly access the data. 
I have added some comments to explain in the code. Let me know if anything is not clear.